src/core/backbone_router/multicast_listeners_table.cpp

```diff
@@ -0,0 +1,318 @@
+/*
+ *  Copyright (c) 2020, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements the Multicast Listeners Table.
+ */
+
+#include "multicast_listeners_table.hpp"
+
+#if OPENTHREAD_FTD && OPENTHREAD_CONFIG_BACKBONE_ROUTER_ENABLE
+
+#include "common/code_utils.hpp"
+#include "common/instance.hpp"
+#include "common/locator-getters.hpp"
+#include "common/logging.hpp"
+#include "common/random.hpp"
+#include "thread/mle_types.hpp"
+#include "thread/thread_netif.hpp"
+#include "thread/thread_uri_paths.hpp"
+
+namespace ot {
+
+namespace BackboneRouter {
+
+otError MulticastListenersTable::Add(const Ip6::Address &aAddress, Time aExpireTime)
+{
+    otError error = OT_ERROR_NONE;
+
+    VerifyOrExit(aAddress.IsMulticastLargerThanRealmLocal(), error = OT_ERROR_INVALID_ARGS);
+
+    for (uint16_t i = 0; i < mNumValidListeners; i++)
+    {
+        Listener &listener = mListeners[i];
+
+        if (listener.GetAddress() == aAddress)
+        {
+            listener.SetExpireTime(aExpireTime);
+            FixHeap(i);
+            ExitNow();
+        }
+    }
+
+    VerifyOrExit(mNumValidListeners < OT_ARRAY_LENGTH(mListeners), error = OT_ERROR_NO_BUFS);
+
+    mListeners[mNumValidListeners].SetAddress(aAddress);
+    mListeners[mNumValidListeners].SetExpireTime(aExpireTime);
+    mNumValidListeners++;
+
+    FixHeap(mNumValidListeners - 1);
```

Comment:
I think it might also work, but the current code make sure all Heap operations operate on a heap with well defined boundaries.   
- `mNumValidListeners` is always the number of elements in the Heap.
- `FixHeap` fixes the element indexed by a given index in the heap. So heap operations actually requires that the operating index  must be valid:
```python3
OT_ASSERT(aIndex < mNumValidListeners);
```

So, in the current code, `FixHeap(mNumValidListeners); mNumValidListeners++` would cause assertion failure.

Reply:
make sense.